File: src/py/sorting.py
"""Sorting and search algorithms in Python."""

from typing import Callable, List, TypeVar

T = TypeVar("T")
Comparator = Callable[[T, T], int]


def _default_compare(a: T, b: T) -> int:
    return (a > b) - (a < b)


def quick_sort(arr: List[T], cmp: Comparator[T] = _default_compare) -> List[T]:
    if len(arr) <= 1:
        return arr[:]
    pivot = arr[len(arr) // 2]
    return (
        quick_sort([x for x in arr if cmp(x, pivot) < 0], cmp)
        + [x for x in arr if cmp(x, pivot) == 0]
        + quick_sort([x for x in arr if cmp(x, pivot) > 0], cmp)
    )


def merge_sort(arr: List[T], cmp: Comparator[T] = _default_compare) -> List[T]:
    if len(arr) <= 1:
        return arr[:]
    mid = len(arr) // 2
    left = merge_sort(arr[:mid], cmp)
    right = merge_sort(arr[mid:], cmp)
    return _merge(left, right, cmp)


def _merge(left: List[T], right: List[T], cmp: Comparator[T]) -> List[T]:
    result: List[T] = []
    i, j = 0, 0
    while i < len(left) and j < len(right):
        if cmp(left[i], right[j]) <= 0:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    return result + left[i:] + right[j:]


def heap_sort(arr: List[T], cmp: Comparator[T] = _default_compare) -> List[T]:
    values, n = arr[:], len(arr)

    def heapify(root: int, size: int) -> None:
        largest = root
        left, right = 2 * root + 1, 2 * root + 2
        if left < size and cmp(values[left], values[largest]) > 0:
            largest = left
        if right < size and cmp(values[right], values[largest]) > 0:
            largest = right
        if largest != root:
            values[root], values[largest] = values[largest], values[root]
            heapify(largest, size)

    for i in range(n // 2 - 1, -1, -1):
        heapify(i, n)
    for i in range(n - 1, 0, -1):
        values[0], values[i] = values[i], values[0]
        heapify(0, i)
    return values


def binary_search(
    arr: List[T], target: T, cmp: Comparator[T] = _default_compare
) -> int:
    low, high = 0, len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        comparison = cmp(arr[mid], target)
        if comparison == 0:
            return mid
        if comparison < 0:
            low = mid + 1
        else:
            high = mid - 1
    return -1


def counting_sort(arr: List[int]) -> List[int]:
    """Sort integers, including negative values, without mutating the input."""
    if not arr:
        return []
    if any(isinstance(value, bool) or not isinstance(value, int) for value in arr):
        raise TypeError("counting_sort accepts integers only")

    min_value = min(arr)
    max_value = max(arr)
    counts = [0] * (max_value - min_value + 1)

    for value in arr:
        counts[value - min_value] += 1

    result: List[int] = []
    for offset, count in enumerate(counts):
        result.extend([offset + min_value] * count)
    return result


def two_sum(nums: List[int], target: int) -> List[int]:
    seen: dict[int, int] = {}
    for index, num in enumerate(nums):
        complement = target - num
        if complement in seen:
            return [seen[complement], index]
        seen[num] = index
    return []

File: src/ts/sorting.ts
export type Comparator<T> = (a: T, b: T) => number;

const defaultComparator = <T>(a: T, b: T): number => (a < b ? -1 : a > b ? 1 : 0);

export function quickSort<T>(arr: readonly T[], cmp: Comparator<T> = defaultComparator): T[] {
  if (arr.length <= 1) {
    return [...arr];
  }

  const pivot = arr[Math.floor(arr.length / 2)];
  const left = arr.filter((value) => cmp(value, pivot) < 0);
  const equal = arr.filter((value) => cmp(value, pivot) === 0);
  const right = arr.filter((value) => cmp(value, pivot) > 0);

  return [...quickSort(left, cmp), ...equal, ...quickSort(right, cmp)];
}

export function mergeSort<T>(arr: readonly T[], cmp: Comparator<T> = defaultComparator): T[] {
  if (arr.length <= 1) {
    return [...arr];
  }

  const midpoint = Math.floor(arr.length / 2);
  return merge(
    mergeSort(arr.slice(0, midpoint), cmp),
    mergeSort(arr.slice(midpoint), cmp),
    cmp,
  );
}

function merge<T>(left: readonly T[], right: readonly T[], cmp: Comparator<T>): T[] {
  const result: T[] = [];
  let leftIndex = 0;
  let rightIndex = 0;

  while (leftIndex < left.length && rightIndex < right.length) {
    const leftValue = left[leftIndex];
    const rightValue = right[rightIndex];

    if (cmp(leftValue, rightValue) <= 0) {
      result.push(leftValue);
      leftIndex += 1;
    } else {
      result.push(rightValue);
      rightIndex += 1;
    }
  }

  return result.concat(left.slice(leftIndex), right.slice(rightIndex));
}

export function heapSort<T>(arr: readonly T[], cmp: Comparator<T> = defaultComparator): T[] {
  const values = [...arr];
  const length = values.length;

  const heapify = (size: number, root: number): void => {
    let largest = root;
    const left = 2 * root + 1;
    const right = 2 * root + 2;

    if (left < size && cmp(values[left], values[largest]) > 0) {
      largest = left;
    }
    if (right < size && cmp(values[right], values[largest]) > 0) {
      largest = right;
    }

    if (largest !== root) {
      [values[root], values[largest]] = [values[largest], values[root]];
      heapify(size, largest);
    }
  };

  for (let index = Math.floor(length / 2) - 1; index >= 0; index -= 1) {
    heapify(length, index);
  }

  for (let index = length - 1; index > 0; index -= 1) {
    [values[0], values[index]] = [values[index], values[0]];
    heapify(index, 0);
  }

  return values;
}

export function binarySearch<T>(
  arr: readonly T[],
  target: T,
  cmp: Comparator<T> = defaultComparator,
): number {
  let low = 0;
  let high = arr.length - 1;

  while (low <= high) {
    const midpoint = Math.floor((low + high) / 2);
    const comparison = cmp(arr[midpoint], target);

    if (comparison === 0) {
      return midpoint;
    }

    if (comparison < 0) {
      low = midpoint + 1;
    } else {
      high = midpoint - 1;
    }
  }

  return -1;
}
